### PR TITLE
Data loader fixes and enhancements for v5.0.3

### LINKED ISSUE #726: Check default value of cansas_version property in CansasReader class (Trac #594)
(Ticket raised on behalf of Lewis) !CansasReader has a property cansas_version which determines the xml header written in the files. This currently seems to default to 1.0 and not 1.1 as you get the old header if you call this class directly. However if you save xml files from the GUI they are saved with the new header! This suggests the default is over-ridden somewhere. Needs investigating.

Migrated from http://trac.sasview.org/ticket/594
```json
{
    "status": "accepted",
    "changetime": "2017-10-27T13:50:15",
    "_ts": "2017-10-27 13:50:15.244021+00:00",
    "description": "(Ticket raised on behalf of Lewis) !CansasReader has a property cansas_version which determines the xml header written in the files. This currently seems to default to 1.0 and not 1.1 as you get the old header if you call this class directly. However if you save xml files from the GUI they are saved with the new header! This suggests the default is over-ridden somewhere. Needs investigating.",
    "reporter": "smk78",
    "cc": "",
    "resolution": "",
    "workpackage": "SasView Bug Fixing",
    "time": "2016-07-22T15:22:18",
    "component": "SasView",
    "summary": "Check default value of cansas_version property in CansasReader class",
    "priority": "major",
    "keywords": "",
    "milestone": "SasView 4.3.0",
    "owner": "tim",
    "type": "task"
}
```

### LINKED ISSUE #1002: Projects should save transmission spectrum (Trac #919)
TO fix issues prior to v4.1.0 release, transmission spectra are no longer saved in project and analysis files. Errors occur when the lengths of lambda and T are different. Along the same lines as outlined in #243 (done for 2D data in master and 1D data in branch load_speed_increase but not merged), we could linearize the arrays into a single <!TransmissionSpectrum> element.

Migrated from http://trac.sasview.org/ticket/919
```json
{
    "status": "new",
    "changetime": "2017-10-26T13:28:15",
    "_ts": "2017-10-26 13:28:15.997491+00:00",
    "description": "TO fix issues prior to v4.1.0 release, transmission spectra are no longer saved in project and analysis files. Errors occur when the lengths of lambda and T are different. Along the same lines as outlined in #14 (done for 2D data in master and 1D data in branch load_speed_increase but not merged), we could linearize the arrays into a single <!TransmissionSpectrum> element.",
    "reporter": "krzywon",
    "cc": "",
    "resolution": "",
    "workpackage": "SasView Bug Fixing",
    "time": "2017-04-04T22:05:42",
    "component": "SasView",
    "summary": "Projects should save transmission spectrum",
    "priority": "minor",
    "keywords": "",
    "milestone": "SasView 5.0.0",
    "owner": "",
    "type": "defect"
}
```

### PR DESCRIPTION
**Pull request by krzywon**
This is a small subset of the changes made in PR #1559 that would be good to include in v5.0.3.

- Fixes #1002: Transmission spectra are now saved when saving in the canSAS XML format. Unit tests were added to verify this is true.
- Fixes #726: The canSAS XML reader now defaults to canSAS v1.1 when reading and writing files.
- Enhancement: SasView version now added as a note when outputting NXcanSAS and canSAS XML files. Unit tests added to verify.
- Enhancement: Updated error handling in the NXcanSAS reader so individual errors do not cause a complete load failure.
- Bug Fix: Fixed an issue when loading scalar quantities in the NXcanSAS reader that predicated the fix above.
- Bug Fix: Fixed an issue loading data with arbitrary units (issue found while loading DAWN-generated data).
- Enhancement: Added unit tests for single-frame and multi-frame NXcanSAS files (these touch the arbitrary unit fix above).

### CODE REVIEW

File: src/sas/sascalc/dataloader/readers/cansas_reader.py [hunk 1]
```diff
@@ -1169,11 +1163,13 @@ def _write_notes(self, datainfo, entry_node):
...
-            self.append(node, entry_node)
-        else:
-            for item in datainfo.notes:
+        from sas.sasview import __version__ as version
+        node = self.create_element("SASnote")
+        self.write_text(
+            node, 'Data file generated by SasView v{0}'.format(version))
+        self.append(node, entry_node)
+        for item in datainfo.notes:
+            if item:
```
Comment (pkienzle) [inline src/sas/sascalc/dataloader/readers/cansas_reader.py:1172]: Do you need the `if item` here?  That is, can  `datainfo.notes` include an empty note?

File: src/sas/sascalc/dataloader/readers/cansas_reader.py [hunk 2]
```diff
@@ -1169,11 +1163,13 @@ def _write_notes(self, datainfo, entry_node):
...
         """
-        if len(datainfo.notes) == 0:
-            node = self.create_element("SASnote")
-            self.append(node, entry_node)
-        else:
-            for item in datainfo.notes:
+        from sas.sasview import __version__ as version
+        node = self.create_element("SASnote")
+        self.write_text(
+            node, 'Data file generated by SasView v{0}'.format(version))
```
Comment (pkienzle) [inline src/sas/sascalc/dataloader/readers/cansas_reader.py:1169]: Thinking toward splitting sascalc into its own repo, if another program is using sascalc to write canSAS data it is not correct to say that the data is generated by SasView. You will need a different method to indicate how the data was generated, such as adding it to datainfo in the GUI or adding `notes=None` to the `write` method.

If you stick with the current scheme, any reason not to put the import at the top of the file?

File: src/sas/sascalc/file_converter/nxcansas_writer.py [hunk 3]
```diff
@@ -286,16 +286,20 @@ def _write_h5_vector(entry, vector, names=['x_position', 'y_position'],
...
-        note_entry = sasentry.create_group('sasnote'.format(i))
-        note_entry.attrs['canSAS_class'] = 'SASnote'
-        notes = None
-        if len(data_info.notes) > 1:
-            notes = [np.string_(n) for n in data_info.notes]
-            notes = np.array(notes)
-        elif data_info.notes != []:
-            notes = _h5_string(data_info.notes[0])
+        from sas.sasview import __version__ as VERSION
+        writer_note = 'Data file generated by SasView v{0}'.format(VERSION)
```
[the remark this answers is not in the document]
Comment (krzywon) [reply src/sas/sascalc/file_converter/nxcansas_writer.py:290]: These are two different writers - one for canSAS XML (in dataloader) and one for NXcanSAS (in fileconverter). Long term (or here if I get it working) these should be in the same place.

File: src/sas/sascalc/dataloader/readers/cansas_reader.py (hunk 1 above)
Comment (krzywon) [reply src/sas/sascalc/dataloader/readers/cansas_reader.py:1172]: Oddly enough, yes, they can be.

File: src/sas/sascalc/dataloader/readers/cansas_reader.py (hunk 2 above)
Comment (krzywon) [reply src/sas/sascalc/dataloader/readers/cansas_reader.py:1169]: I'll look into changing the write method for this PR, but if that doesn't happen, I will add that to the larger loader PR.

File: src/sas/sascalc/file_converter/nxcansas_writer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (krzywon) [reply src/sas/sascalc/file_converter/nxcansas_writer.py]: I wasn't able to get both writers in the same place. There is a separate canSAS XML writer for the file converter and I'm still trying to understand why.

File: src/sas/sascalc/dataloader/readers/cansas_reader_HDF5.py [hunk 4]
```diff
@@ -2,11 +2,12 @@
     NXcanSAS data reader for reading HDF5 formatted CanSAS files.
 """
 
+import logging
 import h5py
 import numpy as np
 import re
 import os
-import sys
+import traceback
```
Comment (pkienzle) [inline src/sas/sascalc/dataloader/readers/cansas_reader_HDF5.py:10]: pep8 suggests imports standard library imports come before third party packages, so h5py and numpy should drop below, with a blank line before.

File: src/sas/sascalc/file_converter/nxcansas_writer.py [hunk 5]
```diff
@@ -286,16 +286,20 @@ def _write_h5_vector(entry, vector, names=['x_position', 'y_position'],
...
+            notes = [np.string_(data_info.notes)]
+        elif len(data_info.notes) > 1:
+            # Handle iterables that aren't empty
             notes = [np.string_(n) for n in data_info.notes]
-            notes = np.array(notes)
-        elif data_info.notes != []:
-            notes = _h5_string(data_info.notes[0])
+        else:
+            # Notes are required in NXcanSAS format
+            notes = [np.string_('')]
```
Comment (pkienzle) [inline src/sas/sascalc/file_converter/nxcansas_writer.py:297]: `np.string_` is [almost undocumented](https://numpy.org/devdocs/reference/routines.char.html).  It has the same docstring as `bytes` but it is not equal to `bytes` nor is it a subclass of `bytes`.  I suspect it is a holdover from 2-3 conversion, and you could `bytes` directly, or in this case, `b''`.

**Pull request merged**